python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -554,7 +601,12 @@ def reconcile(
         Tuple[Optional[PartitionsDefinition], Optional[str]], Set[AssetKey]
     ] = defaultdict(set)
 
-    for asset_partition in asset_partitions_to_reconcile:
+    for asset_partition in (
+        asset_partitions_to_reconcile | asset_partitions_to_reconcile_for_freshness
+    ):
+        # defer materializing the asset now if it will be materialized later
+        if asset_partition in eventual_asset_partitions_to_reconcile_for_freshness:
```

Comment:
Imagine we have 

```
       /-> C
A -> B
       \-> D
```

C has a freshness policy and D does not. `asset_partitions_to_reconcile` includes B, C, and D, and `eventual_asset_partitions_to_reconcile_for_freshness ` includes B and C.  So we defer B and C, but we still execute D. Which is problematic, because D shouldn't execute until B executes.

I wonder if the freshness algorithm should be influencing what shows up in `to_reconcile` / `candidates_queue` in `determine_asset_partitions_to_reconcile`?

Reply:
good catch, I *think* this can be solved inside of determine_asset_partitions_to_reconcile. If we add an additional condition to allow an assetkeypartitionkey to be added to the `to_reconcile` set, which is that it cannot be part of the eventual_asset_partitions_to_reconcile_for_freshness set, then this should stop changes from propagating downstream (i.e. B would not be added to the set, so D would not have all of its parents in to_reconcile, so it would not be added to the set either)